File: app/src/main/java/com/druger/aboutwork/fragments/SelectedReviewFragment.java
package com.druger.aboutwork.fragments;

import android.content.res.Resources;
import android.graphics.Color;
import android.os.Bundle;
import android.text.Editable;
import android.text.TextWatcher;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.EditText;
import android.widget.ImageView;
import android.widget.RatingBar;
import android.widget.TextView;

import androidx.annotation.Nullable;
import androidx.appcompat.app.AlertDialog;
import androidx.constraintlayout.widget.Group;
import androidx.fragment.app.FragmentTransaction;
import androidx.recyclerview.widget.DefaultItemAnimator;
import androidx.recyclerview.widget.LinearLayoutManager;
import androidx.recyclerview.widget.RecyclerView;

import com.druger.aboutwork.R;
import com.druger.aboutwork.activities.MainActivity;
import com.druger.aboutwork.adapters.CommentAdapter;
import com.druger.aboutwork.db.FirebaseHelper;
import com.druger.aboutwork.interfaces.OnItemClickListener;
import com.druger.aboutwork.interfaces.view.SelectedReview;
import com.druger.aboutwork.model.Comment;
import com.druger.aboutwork.model.Review;
import com.druger.aboutwork.presenters.SelectedReviewPresenter;
import com.druger.aboutwork.utils.Utils;
import com.google.firebase.auth.FirebaseUser;

import org.threeten.bp.Instant;
import org.threeten.bp.LocalDate;
import org.threeten.bp.Period;
import org.threeten.bp.ZoneId;

import java.util.Calendar;
import java.util.Date;
import java.util.List;

import moxy.presenter.InjectPresenter;

import static com.druger.aboutwork.Const.Bundles.EDIT_MODE;
import static com.druger.aboutwork.Const.Colors.DISLIKE;
import static com.druger.aboutwork.Const.Colors.GRAY_500;
import static com.druger.aboutwork.Const.Colors.LIKE;
import static com.druger.aboutwork.Const.Colors.PURPLE_100;
import static com.druger.aboutwork.Const.Colors.PURPLE_500;

public class SelectedReviewFragment extends BaseSupportFragment implements View.OnClickListener, SelectedReview {
    private static final int NEW = 0;
    private static final int UPDATE = 1;
    private static final String REVIEW_KEY = "review_key";

    private int type = NEW;

    @InjectPresenter
    SelectedReviewPresenter presenter;

    private TextView tvDescriptionStatus;
    private TextView tvStatus;
    private ImageView ivLike;
    private ImageView ivDislike;
    private Review review;
    private String reviewKey;
    private ImageView ivEdit;
    private EditText etMessage;
    private ImageView ivSend;
    private TextView tvPluses;
    private TextView tvMinuses;
    private TextView tvName;
    private TextView tvDate;
    private TextView tvPosition;
    private RatingBar rbSalary;
    private RatingBar rbCareer;
    private RatingBar rbCollective;
    private RatingBar rbSocialPackage;
    private RatingBar rbChief;
    private RatingBar rbWorkplace;
    private TextView tvDislike;
    private TextView tvLike;
    private RecyclerView rvComments;
    private CommentAdapter commentAdapter;
    private Group groupRating;

    private Bundle bundle;
    private boolean editMode;

    public static SelectedReviewFragment newInstance(String reviewKey, boolean editMode) {

        Bundle args = new Bundle();
        args.putString(REVIEW_KEY, reviewKey);
        args.putBoolean(EDIT_MODE, editMode);

        SelectedReviewFragment fragment = new SelectedReviewFragment();
        fragment.setArguments(args);
        return fragment;
    }

    @Override
    public View onCreateView(LayoutInflater inflater, ViewGroup container,
                             Bundle savedInstanceState) {
        getBundles();
        setView(inflater, container);
        setupToolbar();
        setUI();
        setUX();
        getReview();
        return rootView;
    }

    private void getBundles() {
        bundle = getArguments();
        reviewKey = bundle.getString(REVIEW_KEY);
        editMode = getArguments().getBoolean(EDIT_MODE);
    }

    @Override
    public void setupComments(@Nullable FirebaseUser user) {
        commentAdapter = new CommentAdapter(presenter.getUser());
        rvComments.setLayoutManager(new LinearLayoutManager(getActivity()));
        rvComments.setItemAnimator(new DefaultItemAnimator());
        rvComments.setAdapter(commentAdapter);
        commentAdapter.setOnNameClickListener(comment -> showReviews(comment.getUserId()));

        retrieveComments();
        setupListeners();
    }

    private void retrieveComments() {
        presenter.retrieveComments(reviewKey);
    }

    private void setupListeners() {
        etMessage.addTextChangedListener(new TextWatcher() {
            @Override
            public void beforeTextChanged(CharSequence s, int start, int count, int after) {

            }

            @Override
            public void onTextChanged(CharSequence s, int start, int before, int count) {
                if (s.toString().trim().length() > 0) {
                    ivSend.setClickable(true);
                    ivSend.setColorFilter(Color.parseColor(PURPLE_500));
                } else {
                    ivSend.setClickable(false);
                    ivSend.setColorFilter(Color.parseColor(PURPLE_100));
                }
            }

            @Override
            public void afterTextChanged(Editable s) {

            }
        });

        ivSend.setOnClickListener(v -> {
            switch (type) {
                case NEW:
                    sendMessage(etMessage.getText().toString().trim(), NEW);
                    break;
                case UPDATE:
                    sendMessage(etMessage.getText().toString().trim(), UPDATE);
                    break;
            }
        });

        commentAdapter.setOnItemClickListener(new OnItemClickListener<Comment>() {
            @Override
            public void onClick(Comment item, int position) {

            }

            @Override
            public boolean onLongClick(int position) {
                return presenter.onLongClick(position);
            }
        });

    }

    private void sendMessage(String message, int type) {
        if (message.length() > 0) {
            if (type == NEW) {
                presenter.addComment(message, review.getFirebaseKey());
            } else if (type == UPDATE) {
                presenter.updateComment(message);
                Utils.INSTANCE.hideKeyboard(getActivity(), this.etMessage);
                this.type = NEW;
            }
            this.etMessage.setText(null);
        }
    }

    private void setView(LayoutInflater inflater, ViewGroup container) {
        rootView = inflater.inflate(R.layout.fragment_selected_review, container, false);
        if (editMode) {
            ((MainActivity) getActivity()).hideBottomNavigation();
        }
    }

    private void setUX() {
        ivLike.setOnClickListener(this);
        ivDislike.setOnClickListener(this);
        if (editMode) {
            ivEdit.setOnClickListener(this);
        }
    }

    private void setUI() {
        tvDescriptionStatus = bindView(R.id.tvDescriptionStatus);
        tvStatus = bindView(R.id.tvStatus);
        ivLike = bindView(R.id.ivLike);
        ivDislike = bindView(R.id.ivDislike);
        etMessage = bindView(R.id.etMessage);
        ivSend = bindView(R.id.ivSend);
        rvComments = bindView(R.id.rvComments);
        tvPluses = bindView(R.id.tvPluses);
        tvMinuses = bindView(R.id.tvMinuses);
        tvName = bindView(R.id.tvName);
        tvDate = bindView(R.id.tvDate);
        tvPosition = bindView(R.id.tvPosition);
        rbSalary = bindView(R.id.ratingbar_salary);
        rbCareer = bindView(R.id.ratingbar_career);
        rbCollective = bindView(R.id.ratingbar_collective);
        rbSocialPackage = bindView(R.id.ratingbar_social_package);
        rbChief = bindView(R.id.ratingbar_chief);
        rbWorkplace = bindView(R.id.ratingbar_workplace);
        tvDislike = bindView(R.id.tvDislike);
        tvLike = bindView(R.id.tvLike);
        groupRating = bindView(R.id.group_rating);
    }

    private void setupToolbar() {
        mToolbar = bindView(R.id.toolbar);
        ivEdit = bindView(R.id.ivEdit);
        setActionBar(mToolbar);
        getActionBar().setDisplayHomeAsUpEnabled(true);
        getActionBar().setTitle(R.string.review);
        if (editMode) ivEdit.setVisibility(View.VISIBLE);
         else ivEdit.setVisibility(View.GONE);
    }

    private void getReview() {
        presenter.getReview(reviewKey);
    }

    @Override
    public void setReview(Review review) {
        if (review != null) {
            this.review = review;
            boolean myLike = review.getMyLike();
            boolean myDislike = review.getMyDislike();
            if (!myLike) {
                ivLike.setTag(getActivity().getString(R.string.like_inactive));
            } else {
                ivLike.setTag(getActivity().getString(R.string.like_active));
                ivLike.setColorFilter(Color.parseColor(LIKE));
            }
            if (!myDislike) {
                ivDislike.setTag(getActivity().getString(R.string.dislike_inactive));
            } else {
                ivDislike.setTag(getActivity().getString(R.string.dislike_active));
                ivDislike.setColorFilter(Color.parseColor(DISLIKE));
            }
            tvPluses.setText(Utils.getQuoteSpan(getActivity(), review.getPluses(), R.color.review_positive));
            tvMinuses.setText(Utils.getQuoteSpan(getActivity(), review.getMinuses(), R.color.review_negative));
            tvName.setText(review.getName());
            tvDate.setText(Utils.getDate(review.getDate()));
            tvPosition.setText(review.getPosition());
            rbSalary.setRating(review.getMarkCompany().getSalary());
            rbCareer.setRating(review.getMarkCompany().getCareer());
            rbCollective.setRating(review.getMarkCompany().getCollective());
            rbSocialPackage.setRating(review.getMarkCompany().getSocialPackage());
            rbChief.setRating(review.getMarkCompany().getChief());
            rbWorkplace.setRating(review.getMarkCompany().getWorkplace());
            tvDislike.setText(String.valueOf(review.getDislike()));
            tvLike.setText(String.valueOf(review.getLike()));
            setExperience(review);
        }
    }

    private void setExperience(Review review) {
        switch (review.getStatus()) {
            case Review.WORKING:
                tvStatus.setText(R.string.working);
                setWorkingDays(review.getEmploymentDate(), Calendar.getInstance().getTimeInMillis());
                break;
            case Review.WORKED:
                tvStatus.setText(R.string.worked);
                setWorkingDays(review.getEmploymentDate(), review.getDismissalDate());
                break;
            case Review.INTERVIEW:
                tvStatus.setText(R.string.interview);
                long interviewDate = review.getInterviewDate();
                if (interviewDate != 0) tvDescriptionStatus.setText(Utils.getDate(interviewDate));
                groupRating.setVisibility(View.GONE);
                break;
        }
    }

    private void setWorkingDays(long first, long last) {
        if (first != 0) {
            Date f = new Date(first);
            Date l = new Date(last);
            LocalDate firstDate = Instant.ofEpochMilli(f.getTime())
                    .atZone(ZoneId.systemDefault()).toLocalDate();
            LocalDate lastDate = Instant.ofEpochMilli(l.getTime())
                    .atZone(ZoneId.systemDefault()).toLocalDate();

            Period period = Period.between(firstDate, lastDate);
            int years = period.getYears();
            int months = period.getMonths();
            int days = period.getDays();
            Resources res = getResources();
            if (years > 0 && months > 0 && days > 0) {
                tvDescriptionStatus.setText(res.getQuantityString(R.plurals.year, years, years));
                tvDescriptionStatus.append(" " + res.getQuantityString(R.plurals.month, months,months));
                tvDescriptionStatus.append(" " + res.getQuantityString(R.plurals.day, days, days));
            } else if (years >= 0 && months <= 0 && days <= 0) {
                tvDescriptionStatus.setText(res.getQuantityString(R.plurals.year, years, years));
            } else if (years <= 0 && months > 0 && days >= 0) {
                tvDescriptionStatus.setText(res.getQuantityString(R.plurals.month, months, months));
                tvDescriptionStatus.append(" " + res.getQuantityString(R.plurals.day, days, days));
            } else if (years <= 0 && months <= 0) {
                tvDescriptionStatus.setText(res.getQuantityString(R.plurals.day, days, days));
            }
        }
    }

    @Override
    public void onClick(View v) {
        switch (v.getId()) {
            case R.id.ivLike:
                presenter.clickLike();
                break;
            case R.id.ivDislike:
                presenter.clickDislike();
                break;
            case R.id.ivEdit:
                showEditReview();
                break;
            default:
                break;
        }
    }

    private void showEditReview() {
        EditReviewFragment reviewFragment = EditReviewFragment.Companion.newInstance(reviewKey);
        replaceFragment(reviewFragment, R.id.main_container, true);
    }

    @Override
    public void onLikeClicked() {
        int like = review.getLike();
        int dislike = review.getDislike();
        if (!review.getMyLike()) {
            ivLike.setColorFilter(Color.parseColor(LIKE));
            review.setLike(++like);
            tvLike.setText(String.valueOf(like));
            review.setMyLike(true);

            if (review.getMyDislike()) {
                ivDislike.setColorFilter(Color.parseColor(GRAY_500));
                review.setDislike(--dislike);
                tvDislike.setText(String.valueOf(dislike));
                review.setMyDislike(false);
                FirebaseHelper.INSTANCE.dislikeReview(review);
            }
        } else {
            ivLike.setColorFilter(Color.parseColor(GRAY_500));
            review.setLike(--like);
            tvLike.setText(String.valueOf(like));
            review.setMyLike(false);
        }
        FirebaseHelper.INSTANCE.likeReview(review);
    }

    @Override
    public void onDislikeClicked() {
        int like = review.getLike();
        int dislike = review.getDislike();
        if (!review.getMyDislike()) {
            ivDislike.setColorFilter(Color.parseColor(DISLIKE));
            review.setDislike(++dislike);
            tvDislike.setText(String.valueOf(dislike));
            review.setMyDislike(true);

            if (review.getMyLike()) {
                ivLike.setColorFilter(Color.parseColor(GRAY_500));
                review.setLike(--like);
                tvLike.setText(String.valueOf(like));
                review.setMyLike(false);
                FirebaseHelper.INSTANCE.likeReview(review);
            }
        } else {
            ivDislike.setColorFilter(Color.parseColor(GRAY_500));
            review.setDislike(--dislike);
            tvDislike.setText(String.valueOf(dislike));
            review.setMyDislike(false);
        }
        FirebaseHelper.INSTANCE.dislikeReview(review);
    }

    private void showReviews(String userId) {
        UserReviewsFragment reviews = UserReviewsFragment.Companion.newInstance(userId);

        FragmentTransaction transaction = getFragmentManager().beginTransaction();
        transaction.replace(R.id.main_container, reviews);
        transaction.addToBackStack(null);
        transaction.commit();
    }

    @Override
    public void onDestroyView() {
        super.onDestroyView();
        if (editMode) {
            ((MainActivity) getActivity()).showBottomNavigation();
        }
    }

    @Override
    public void onDestroy() {
        super.onDestroy();
        presenter.removeListeners();
    }

    @Override
    public void showChangeDialog(final int position) {
        AlertDialog.Builder builder = new AlertDialog.Builder(getActivity());
        builder.setItems(R.array.comments_change, (dialog, which) -> {
            switch (which) {
                case 0:
                    presenter.deleteComment(position);
                    break;
                case 1:
                    etMessage.setText(presenter.getComment().getMessage());
                    Utils.INSTANCE.showKeyboard(getActivity());
                    etMessage.setFocusableInTouchMode(true);
                    etMessage.setSelection(presenter.getComment().getMessage().length());
                    type = UPDATE;
                    break;
            }
        });
        builder.show();
    }

    @Override
    public void notifyItemRemoved(int position, int size) {
        commentAdapter.notifyItemRemoved(position);
        commentAdapter.notifyItemRangeChanged(position, size);
    }

    @Override
    public void showComments(List<Comment> comments) {
        commentAdapter.clear();
        commentAdapter.addItems(comments);
    }

    @Override
    public void showAuthDialog(int title) {
        Utils.INSTANCE.showAuthDialog(getActivity(), title);
    }
}

File: app/src/main/java/com/druger/aboutwork/fragments/SelectedReviewFragment.kt
package com.druger.aboutwork.fragments

import android.graphics.Color
import android.os.Bundle
import android.text.Editable
import android.text.TextWatcher
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import androidx.appcompat.app.AlertDialog
import androidx.collection.ArrayMap
import androidx.core.content.ContextCompat
import androidx.core.content.res.ResourcesCompat
import androidx.core.view.isVisible
import androidx.fragment.app.viewModels
import androidx.recyclerview.widget.DefaultItemAnimator
import com.druger.aboutwork.R
import com.druger.aboutwork.activities.MainActivity
import com.druger.aboutwork.adapters.CommentAdapter
import com.druger.aboutwork.adapters.PhotoAdapter
import com.druger.aboutwork.databinding.FragmentSelectedReviewBinding
import com.druger.aboutwork.db.FirebaseHelper
import com.druger.aboutwork.enums.Screen
import com.druger.aboutwork.fragments.ReviewFragment.Companion.CURRENT_PHOTO_POSITION
import com.druger.aboutwork.fragments.ReviewFragment.Companion.FULL_SCREEN_STORAGE
import com.druger.aboutwork.interfaces.OnItemClickListener
import com.druger.aboutwork.model.Comment
import com.druger.aboutwork.model.Review
import com.druger.aboutwork.utils.Utils
import com.druger.aboutwork.viewmodels.SelectedReviewViewModel
import com.google.android.material.transition.MaterialContainerTransform
import com.google.firebase.auth.FirebaseUser
import com.google.firebase.storage.StorageReference
import dagger.hilt.android.AndroidEntryPoint
import org.threeten.bp.Instant
import org.threeten.bp.Period
import org.threeten.bp.ZoneId
import java.util.Calendar
import java.util.Date

@AndroidEntryPoint
class SelectedReviewFragment : BaseSupportFragment() {

    private var type = NEW

    private val viewModel: SelectedReviewViewModel by viewModels()

    private var _binding: FragmentSelectedReviewBinding? = null
    private val binding get() = _binding!!

    private lateinit var commentAdapter: CommentAdapter
    private lateinit var photoAdapter: PhotoAdapter<StorageReference>
    private var isFullScreenShown = false
    private var currentPhotoPosition = 0

    private var review: Review? = null
    private var reviewKey: String? = null
    private var message: String? = null // if we came after Login shouldn't be null
    private var editMode: Boolean = false
    private var showUserName: Boolean = false
    private var likesDislikes: MutableMap<String, Boolean>? = null

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        sharedElementEnterTransition = MaterialContainerTransform().apply {
            startContainerColor = ContextCompat.getColor(requireContext(), R.color.colorSurface)
            endContainerColor = ContextCompat.getColor(requireContext(), R.color.colorSurface)
            scrimColor = Color.TRANSPARENT
        }
        photoAdapter = PhotoAdapter(mutableListOf(), false)
        observeClearMessage()
        observeAuth()
        observeChangeDialog()
        observeDeleteComment()
        observeComments()
        observeReview()
        observeLikeClick()
        observeDislikeClick()
        observeUserReviews()
        observeCompanyDetails()
        observePhotos()
    }

    private fun observePhotos() {
        viewModel.photosState.observe(this) { showPhotos(it) }
    }

    private fun observeCompanyDetails() {
        viewModel.companyDetailsState.observe(this) { showCompanyDetail(it) }
    }

    private fun observeUserReviews() {
        viewModel.userReviewsState.observe(this) { showUserReviews(it) }
    }

    private fun observeDislikeClick() {
        viewModel.dislikeClickState.observe(this) { onDislikeClicked() }
    }

    private fun observeLikeClick() {
        viewModel.likeClickState.observe(this) { onLikeClicked() }
    }

    private fun observeReview() {
        viewModel.reviewState.observe(this) { setReview(it) }
    }

    private fun observeComments() {
        viewModel.commentsState.observe(this) { showComments(it) }
    }

    private fun observeDeleteComment() {
        viewModel.deleteCommentState.observe(this) { position ->
            notifyItemRemoved(position, viewModel.comments.size)
        }
    }

    private fun observeChangeDialog() {
        viewModel.changeDialogState.observe(this) { showChangeDialog(it) }
    }

    private fun observeAuth() {
        viewModel.authState.observe(this) { showAuth(it) }
    }

    private fun observeClearMessage() {
        viewModel.clearMessageState.observe(this) { clearMessage() }
    }

    override fun onCreateView(
        inflater: LayoutInflater, container: ViewGroup?,
        savedInstanceState: Bundle?
    ): View {
        getBundles()
        _binding = FragmentSelectedReviewBinding.inflate(inflater, container, false)
        (activity as MainActivity).hideBottomNavigation()
        getReview()
        setupComments(viewModel.user)
        return binding.root
    }

    override fun onViewCreated(view: View, savedInstanceState: Bundle?) {
        super.onViewCreated(view, savedInstanceState)
        setupToolbar()
        setUX()
    }

    private fun getBundles() {
        arguments?.apply {
            reviewKey = getString(REVIEW_KEY)
            message = getString(MESSAGE)
            editMode = getBoolean(EDIT_MODE)
            showUserName = getBoolean(SHOW_USER_NAME)
        }
    }

    override fun onViewStateRestored(savedInstanceState: Bundle?) {
        super.onViewStateRestored(savedInstanceState)
        if (savedInstanceState != null) {
            currentPhotoPosition = savedInstanceState.getInt(CURRENT_PHOTO_POSITION)
            isFullScreenShown = savedInstanceState.getBoolean(FULL_SCREEN_STORAGE)
        }
        if (isFullScreenShown) photoAdapter.showFullScreen(
            requireContext(),
            currentPhotoPosition,
            null
        )
    }

    override fun onSaveInstanceState(outState: Bundle) {
        outState.putInt(CURRENT_PHOTO_POSITION, photoAdapter.currentPosition)
        outState.putBoolean(FULL_SCREEN_STORAGE, photoAdapter.isFullScreen)
        super.onSaveInstanceState(outState)
    }

    private fun setupComments(user: FirebaseUser?) {
        commentAdapter = CommentAdapter(user)
        with(binding.ltContent.rvComments) {
            itemAnimator = DefaultItemAnimator()
            adapter = commentAdapter
        }
        commentAdapter.setOnNameClickListener(object : CommentAdapter.OnNameClickListener {
            override fun onClick(comment: Comment) {
                showReviews(comment.userId)
            }
        })

        retrieveComments()
        setupListeners()
    }

    private fun retrieveComments() {
        reviewKey?.let { viewModel.retrieveComments(it) }
    }

    private fun setupListeners() {
        with(binding.ltContent) {
            etMessage.addTextChangedListener(object : TextWatcher {
                override fun beforeTextChanged(
                    s: CharSequence,
                    start: Int,
                    count: Int,
                    after: Int
                ) {
                }

                override fun onTextChanged(s: CharSequence, start: Int, before: Int, count: Int) {
                    if (s.toString().trim().isNotEmpty()) {
                        ivSend.isClickable = true
                        ivSend.setColorFilter(
                            ResourcesCompat.getColor(
                                resources, R.color.colorPrimary, null
                            )
                        )
                    } else {
                        ivSend.isClickable = false
                        ivSend.setColorFilter(
                            ResourcesCompat.getColor(
                                resources, R.color.colorPrimaryLight, null
                            )
                        )
                    }
                }

                override fun afterTextChanged(s: Editable) {}
            })

            ivSend.setOnClickListener {
                when (type) {
                    NEW -> sendMessage(etMessage.text.toString().trim(), NEW)
                    UPDATE -> sendMessage(etMessage.text.toString().trim(), UPDATE)
                }
            }
        }

        commentAdapter.setOnItemClickListener(object : OnItemClickListener<Comment> {
            override fun onClick(item: Comment, position: Int) {}

            override fun onLongClick(item: Comment, position: Int): Boolean {
                return viewModel.onLongClick(position)
            }
        })
    }

    private fun sendMessage(message: String, type: Int) {
        if (message.isNotEmpty()) {
            if (type == NEW) {
                review?.firebaseKey?.let { viewModel.addComment(message, it) }
            } else if (type == UPDATE) {
                viewModel.updateComment(message)
                Utils.hideKeyboard(requireContext(), binding.ltContent.etMessage)
                this.type = NEW
            }
        }
    }

    private fun setUX() {
        with(binding.ltContent) {
            ivLike.setOnClickListener { viewModel.clickLike() }
            ivDislike.setOnClickListener { viewModel.clickDislike() }
            if (editMode) {
                binding.toolbar.ivEdit.setOnClickListener { showEditReview() }
            }
            tvName.setOnClickListener { viewModel.onClickName(showUserName) }
        }
    }

    private fun setupToolbar() {
        with(binding.toolbar) {
            setActionBar(toolbar)
            actionBar?.setDisplayHomeAsUpEnabled(true)
            actionBar?.setTitle(R.string.review)
            ivEdit.isVisible = editMode
        }
    }

    private fun getReview() {
        reviewKey?.let { viewModel.getReview(it, showUserName) }
    }

    private fun setReview(review: Review?) {
        review?.let {
            this.review = review
            setMyLikeDislike(review)
            with(binding.ltContent) {
                tvPluses.text =
                    Utils.getQuoteSpan(requireContext(), review.pluses, R.color.review_positive)
                tvMinuses.text =
                    Utils.getQuoteSpan(requireContext(), review.minuses, R.color.review_negative)
                tvName.text = review.name
                tvDate.text = Utils.getDate(review.date)
                tvPosition.text = review.position
                tvDislike.text = review.dislike.toString()
                tvLike.text = review.like.toString()
                message?.let { etMessage.setText(it) }
            }
            setMarkCompany(review)
            setExperience(review)
            checkMessage()
            setRecommendation(review)
            viewModel.getPhotos(review.firebaseKey)
        }
    }

    private fun setMyLikeDislike(review: Review) {
        with(binding.ltContent) {
            ivLike.setColorFilter(
                ResourcesCompat.getColor(
                    requireContext().resources, R.color.like_disable, null
                )
            )
            ivDislike.setColorFilter(
                ResourcesCompat.getColor(
                    requireContext().resources, R.color.like_disable, null
                )
            )
            likesDislikes = review.likesDislikes
            likesDislikes?.let { likes ->
                likes[viewModel.user?.uid]?.let { myLike ->
                    if (myLike) {
                        ivLike.setColorFilter(
                            ResourcesCompat.getColor(
                                requireContext().resources, R.color.like, null
                            )
                        )
                    } else {
                        ivDislike.setColorFilter(
                            ResourcesCompat.getColor(
                                requireContext().resources, R.color.dislike, null
                            )
                        )
                    }
                }
            }
        }
    }

    private fun setMarkCompany(review: Review) {
        with(binding.ltContent.markCompany) {
            review.markCompany?.let { mark ->
                rbSalary.rating = mark.salary
                rbCareer.rating = mark.career
                rbCollective.rating = mark.collective
                rbSocialPackage.rating = mark.socialPackage
                rbChief.rating = mark.chief
                rbWorkplace.rating = mark.workplace
            }
        }
    }

    private fun setRecommendation(review: Review) {
        with(binding.ltContent) {
            review.recommended?.let { recommended ->
                ivRecommendation.isVisible = true
                tvRecommendation.isVisible = true
                if (recommended) {
                    ivRecommendation.setImageResource(R.drawable.ic_recommended)
                    tvRecommendation.text = getString(R.string.recommended)
                } else {
                    ivRecommendation.setImageResource(R.drawable.ic_not_recommended)
                    tvRecommendation.text = getString(R.string.not_recommended)
                }
            }
        }
    }

    private fun checkMessage() {
        message?.let { sendMessage(it, NEW) }
    }

    private fun setExperience(review: Review) {
        with(binding.ltContent) {
            when (review.status) {
                Review.WORKING -> {
                    tvStatus.setText(R.string.working)
                    setWorkingDays(review.employmentDate, Calendar.getInstance().timeInMillis)
                }

                Review.WORKED -> {
                    tvStatus.setText(R.string.worked)
                    setWorkingDays(review.employmentDate, review.dismissalDate)
                }

                Review.INTERVIEW -> {
                    tvStatus.setText(R.string.interview)
                    markCompany.root.isVisible = false
                }
            }
        }
    }

    private fun setWorkingDays(first: Long, last: Long) {
        if (first != 0L) {
            val f = Date(first)
            val l = Date(last)
            val firstDate = Instant.ofEpochMilli(f.time)
                .atZone(ZoneId.systemDefault()).toLocalDate()
            val lastDate = Instant.ofEpochMilli(l.time)
                .atZone(ZoneId.systemDefault()).toLocalDate()

            val period = Period.between(firstDate, lastDate)
            val years = period.years
            val months = period.months
            val days = period.days
            val res = resources
            with(binding.ltContent) {
                if (years > 0 && months > 0 && days > 0) {
                    tvDescriptionStatus.text = res.getQuantityString(R.plurals.year, years, years)
                    tvDescriptionStatus.append(
                        " " + res.getQuantityString(
                            R.plurals.month,
                            months,
                            months
                        )
                    )
                    tvDescriptionStatus.append(
                        " " + res.getQuantityString(
                            R.plurals.day,
                            days,
                            days
                        )
                    )
                } else if (years >= 0 && months <= 0 && days <= 0) {
                    tvDescriptionStatus.text = res.getQuantityString(R.plurals.year, years, years)
                } else if (years <= 0 && months > 0 && days >= 0) {
                    tvDescriptionStatus.text =
                        res.getQuantityString(R.plurals.month, months, months)
                    tvDescriptionStatus.append(
                        " " + res.getQuantityString(
                            R.plurals.day,
                            days,
                            days
                        )
                    )
                } else if (years <= 0 && months <= 0) {
                    tvDescriptionStatus.text = res.getQuantityString(R.plurals.day, days, days)
                }
            }
        }
    }

    private fun showEditReview() {
        reviewKey?.let {
            val reviewFragment = EditReviewFragment.newInstance(it)
            replaceFragment(reviewFragment, R.id.main_container, true)
        }
    }

    private fun onLikeClicked() {
        review?.let { review ->
            if (likesDislikes == null) likesDislikes = ArrayMap<String, Boolean>()
            var likes = review.like
            var dislikes = review.dislike
            val userId = viewModel.user?.uid
            val myLikeDislike = likesDislikes?.get(userId)
            with(binding.ltContent) {
                myLikeDislike?.let { likeDislike ->
                    if (likeDislike) {
                        review.like = --likes
                        tvLike.text = likes.toString()
                        likesDislikes?.remove(userId)
                    } else {
                        review.dislike = --dislikes
                        tvDislike.text = dislikes.toString()
                        review.like = ++likes
                        tvLike.text = likes.toString()

                        userId?.let { likesDislikes?.put(it, true) }
                    }

                } ?: run {
                    review.like = ++likes
                    tvLike.text = likes.toString()
                    userId?.let { likesDislikes?.put(it, true) }
                }
            }
            review.likesDislikes = this.likesDislikes
            FirebaseHelper.likeOrDislikeReview(review)
        }
    }

    private fun onDislikeClicked() {
        review?.let { review ->
            if (likesDislikes == null) likesDislikes = ArrayMap()
            var likes = review.like
            var dislikes = review.dislike
            val userId = viewModel.user?.uid
            val myLikeDislike = likesDislikes?.get(userId)

            with(binding.ltContent) {
                myLikeDislike?.let { likeDislike ->
                    if (!likeDislike) {
                        review.dislike = --dislikes
                        tvDislike.text = dislikes.toString()
                        likesDislikes?.remove(userId)
                    } else {
                        review.like = --likes
                        tvLike.text = likes.toString()
                        review.dislike = ++dislikes
                        tvDislike.text = dislikes.toString()

                        userId?.let { likesDislikes?.put(it, false) }
                    }

                } ?: run {
                    review.dislike = ++dislikes
                    tvDislike.text = dislikes.toString()
                    userId?.let { likesDislikes?.put(it, false) }
                }
            }
            review.likesDislikes = this.likesDislikes
            FirebaseHelper.likeOrDislikeReview(review)
        }
    }

    private fun showReviews(userId: String?) {
        val reviews = userId?.let { UserReviewsFragment.newInstance(it) }

        fragmentManager?.beginTransaction()?.apply {
            reviews?.let { replace(R.id.main_container, it) }
            addToBackStack(null)
            commit()
        }
    }

    override fun onDestroyView() {
        super.onDestroyView()
        actionBar?.setDisplayHomeAsUpEnabled(false)
    }

    override fun onDestroy() {
        super.onDestroy()
        viewModel.removeListeners()
    }

    private fun showChangeDialog(position: Int) {
        val builder = AlertDialog.Builder(requireContext())
        builder.setItems(R.array.comments_change) { _, which ->
            with(binding.ltContent) {
                when (which) {
                    0 -> viewModel.deleteComment(position)
                    1 -> {
                        etMessage.setText(viewModel.comment.message)
                        Utils.showKeyboard(requireContext())
                        etMessage.isFocusableInTouchMode = true
                        viewModel.comment.message?.length?.let { etMessage.setSelection(it) }
                        type = UPDATE
                    }
                }
            }
        }
        builder.show()
    }

    private fun notifyItemRemoved(position: Int, size: Int) {
        commentAdapter.notifyItemRemoved(position)
        commentAdapter.notifyItemRangeChanged(position, size)
    }

    private fun showComments(comments: List<Comment>) {
        commentAdapter.clear()
        commentAdapter.addItems(comments)
    }

    private fun showAuth(title: Int) {
        with(binding.ltContent) {
            Utils.hideKeyboard(requireContext(), etMessage)
            replaceFragment(
                AuthFragment.newInstance(
                    getString(title),
                    Screen.REVIEW.name,
                    null,
                    reviewKey,
                    etMessage.text.toString().trim()
                ),
                R.id.content_review
            )
        }
    }

    private fun clearMessage() {
        binding.ltContent.etMessage.text = null
    }

    private fun showCompanyDetail(companyId: String?) {
        companyId?.let {
            replaceFragment(CompanyDetailFragment.newInstance(companyId), R.id.main_container, true)
        }
    }

    private fun showUserReviews(userId: String?) {
        userId?.let {
            replaceFragment(
                UserReviewsFragment.newInstance(userId),
                R.id.main_container,
                true
            )
        }
    }

    private fun showPhotos(photos: List<StorageReference>) {
        with(binding.ltContent) {
            rvPhotos.isVisible = true
            photoAdapter.isFullScreen = isFullScreenShown
            photoAdapter.setUri(photos.toMutableList())
            rvPhotos.apply {
                adapter = photoAdapter
                setHasFixedSize(true)
            }
        }
    }

    companion object {
        private const val NEW = 0
        private const val UPDATE = 1
        private const val REVIEW_KEY = "review_key"
        private const val MESSAGE = "message"
        private const val EDIT_MODE = "editMode"
        private const val SHOW_USER_NAME = "show_user_name"

        fun newInstance(
            reviewKey: String,
            editMode: Boolean,
            message: String? = null
        ): SelectedReviewFragment {

            val args = Bundle().apply {
                putString(REVIEW_KEY, reviewKey)
                putString(MESSAGE, message)
                putBoolean(EDIT_MODE, editMode)
            }

            return createSelectedReviewFragment(args)
        }

        fun newInstance(reviewKey: String): SelectedReviewFragment {
            val args = Bundle().apply {
                putString(REVIEW_KEY, reviewKey)
                putBoolean(SHOW_USER_NAME, true)
            }
            return createSelectedReviewFragment(args)
        }

        private fun createSelectedReviewFragment(args: Bundle): SelectedReviewFragment {
            val fragment = SelectedReviewFragment()
            fragment.arguments = args
            return fragment
        }
    }
}
